### PR TITLE
Analyzer: fixes for JOIN columns resolution

### LINKED ISSUE #59067: Experimental analyzer and join_use_nulls are incompatible
join_use_nulls is incompatible with allow_experimental_analyzer.

**Sample query:**
```sql
with d1 as (
    select
        1 as a,
        2 as b
),
d2 as (
    select
        1 as a,
        3 as c
),
joined as (
    select
        d1.*,
        d2.c
    from d1
    inner join d2
        on (d1.a = d2.a)
)
select c 
from joined
SETTINGS 
  allow_experimental_analyzer=1, 
  join_use_nulls=1
```
([fiddle](https://fiddle.clickhouse.com/70767be9-fcab-4251-a4f2-c8e466e3e16d))

**Actual behaviour**
```bash
Received exception from server (version 23.12.2):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Unknown expression identifier 'c' in scope
```

**Expected behaviour**
Returns the number 3.

**Additional info**
 The query works correctly with at most one of the settings being equal to 1, so 
```sql
SETTINGS 
  allow_experimental_analyzer=1, 
  join_use_nulls=0
```
returns 3 as expected, same as the other way around and
```sql
SETTINGS 
  allow_experimental_analyzer=0, 
  join_use_nulls=0
```

**Related issues**
Related to, but is not a duplicate of [this issue](https://github.com/ClickHouse/ClickHouse/issues/58994)

### PR DESCRIPTION
**Pull request by vdimir**
### Changelog category (leave one):

- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

PR contains several fixes (they relates to each other, or found by fuzzer on added test, so it makes sense to merge at once):

* Fixes in support of JOIN USING with ALIAS column (continuation of https://github.com/ClickHouse/ClickHouse/pull/58317):
	* Build actions dag for columns that are not in select list
	* Fixes for multiple JOINs
	* Do not recalculate ALIAS expression in SELECT list (leads to different type)
* Fix for resolving qualified matches with JOIN USING
* Fix for resolving identifier with compound name and join_use_nulls, close https://github.com/ClickHouse/ClickHouse/issues/59067

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit d0889ff71a7213599a85dc04a216f2e1accdb9d5 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59007/d0889ff71a7213599a85dc04a216f2e1accdb9d5/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/59007/d0889ff71a7213599a85dc04a216f2e1accdb9d5/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

File: src/Planner/PlannerJoinTree.cpp [hunk 1]
```diff
@@ -990,40 +990,42 @@ struct UsingAliasKeyActions
...
-        addColumnImpl(left_alias_columns_keys, node, plan_output_columns, planner_context);
-    }
+        for (const auto * input_node : left_alias_columns_keys->getInputs())
+            left_alias_columns_keys->addOrReplaceInOutputs(*input_node);
 
-    void addRightColumn(QueryTreeNodePtr & node, const ColumnsWithTypeAndName & plan_output_columns, const PlannerContextPtr & planner_context)
-    {
-        addColumnImpl(right_alias_columns_keys, node, plan_output_columns, planner_context);
+        for (const auto * input_node : right_alias_columns_keys->getInputs())
+            right_alias_columns_keys->addOrReplaceInOutputs(*input_node);
```
Comment (novikd) [inline src/Planner/PlannerJoinTree.cpp:998]: Isn't it enough to use `projectInput()`?

**Review by novikd — APPROVED** [review]
LGTM

**Comment by vdimir** [comment]
> AST fuzzer (asan) — Logical error: 'Unexpected return type from plus. Expected Int32. Got Nullable'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/59007/173719c92572e674180a386fa80eba9ab8878819/ast_fuzzer__asan_/report.html)
> AST fuzzer (debug) — Logical error: 'Unexpected return type from plus. Expected Int32. Got Nullable'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/59007/173719c92572e674180a386fa80eba9ab8878819/ast_fuzzer__debug_/report.html)
> AST fuzzer (msan) — Logical error: 'Unexpected return type from plus. Expected Int32. Got Nullable'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/59007/173719c92572e674180a386fa80eba9ab8878819/ast_fuzzer__msan_/report.html)
> AST fuzzer (tsan) — Logical error: 'Unexpected return type from plus. Expected Int32. Got Nullable'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/59007/173719c92572e674180a386fa80eba9ab8878819/ast_fuzzer__tsan_/report.html)
> AST fuzzer (ubsan) — Logical error: 'Unexpected return type from plus. Expected Int32. Got Nullable'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/59007/173719c92572e674180a386fa80eba9ab8878819/ast_fuzzer__ubsan_/report.html)

😵‍💫

File: src/Planner/PlannerJoinTree.cpp (hunk 1 above)
Comment (vdimir) [reply src/Planner/PlannerJoinTree.cpp:998]: `projectInput` means 'pass through unknown columns that are not in a DAG', but inputs should be added to output explicitly.

**Comment by Algunenano** [comment]
> AST fuzzer (asan) — Logical error: 'Unexpected return type from plus. Expected Int32. Got Nullable'. [Details](https://s3.amazonaws.com/clickhouse-test-reports/59007/173719c92572e674180a386fa80eba9ab8878819/ast_fuzzer__asan_/report.html)

This is a [newly introduced](https://github.com/ClickHouse/ClickHouse/pull/59379) assertion that verifies that a function returns the type it said it was going to return in getReturnTypeImpl. It might be a corner case I didn't consider or it might be detecting a real bug

**Comment by vdimir** [comment]
Some of issues mentioned above in the first comment were fixed by refactoring ALIAS column computation by moving it right after reading stage. I rebased PR and kept code changes that still required to fix other things and kept tests.

**Pull request merged**